Expose the original status code in StatusCodeReExecuteFeature

amcasey opened the pull request:
# Expose the original status code in StatusCodeReExecuteFeature

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Expose the original status code in `StatusCodeReExecuteFeature`

## Description

Expose the original status code in `StatusCodeReExecuteFeature` so that it can be used in places like `StaticFileOptions.OnPrepareResponse`.

With this change it will be possible to write code like

```csharp
app.UseStaticFiles(new StaticFileOptions
{
    OnPrepareResponse = responseContext =>
    {
        var context = responseContext.Context;
        if (context.Response.StatusCode == StatusCodes.Status200OK &&
            context.Features[typeof(IStatusCodeReExecuteFeature)] is IStatusCodeReExecuteFeature statusCodeFeature)
        {
            context.Response.StatusCode = statusCodeFeature.OriginalStatusCode;
        }
    }
});
```

Fixes #46538

amcasey committed:
Expose the original status code in StatusCodeReExecuteFeature

...so that it can be used in places like `StaticFileOptions.OnPrepareResponse`.

With this change it will be possible to write code like

```csharp
app.UseStaticFiles(new StaticFileOptions
{
    OnPrepareResponse = responseContext =>
    {
        var context = responseContext.Context;
        if (context.Response.StatusCode == StatusCodes.Status200OK &&
            context.Features[typeof(IStatusCodeReExecuteFeature)] is IStatusCodeReExecuteFeature statusCodeFeature)
        {
            context.Response.StatusCode = statusCodeFeature.OriginalStatusCode;
        }
    }
});
```

Fixes #46538

Tratcher commented on src/Middleware/Diagnostics.Abstractions/src/IStatusCodeReExecuteFeature.cs:
```diff
@@ -28,6 +28,11 @@ public interface IStatusCodeReExecuteFeature
+    /// <summary>
+    /// The <see cref="HttpResponse.StatusCode"/> of the original response.
+    /// </summary>
+    int OriginalStatusCode { get; set; }
```
Can this be done in a non-breaking way with DIM's? Not sure how to do that with properties, but the doc claims it's possible.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods
```suggestion
    public int OriginalStatusCode { get; set; } = StatusCodes.Status200Ok;
```

amcasey replied:
Would it be better to make it nullable?  200 would be confusing in scenarios where the original status code was something else.

amcasey replied:
I guess you could use -1 or something if you were _really_ opposed to boxing.

amcasey replied:
Also, do people really implement this interface?  How would they use their own implementations?

amcasey replied:
I've updated the API proposal.

amcasey replied:
We decided to make it throw.

Tratcher reviewed (commented):
Tests?

amcasey commented:
Force push is a trivial rebase.  API Review changes to follow.

amcasey committed:
Drop setter and make default getter throw

amcasey commented:
> Tests?

There aren't any for `IStatusCodeReExecuteFeature`, AFAICT.

amcasey committed:
Remove obsolete public API stub

amcasey commented on src/Middleware/Diagnostics/src/StatusCodePage/StatusCodeReExecuteFeature.cs:
```diff
@@ -18,6 +18,9 @@ public class StatusCodeReExecuteFeature : IStatusCodeReExecuteFeature
     public string? OriginalQueryString { get; set; }
 
+    /// <inheritdoc/>
+    public int OriginalStatusCode { get; set; }
```
```suggestion
    public int OriginalStatusCode { get; internal set; }
```

amcasey committed:
Make setter internal

...the other setters in the type should also be internal, but it's too late now.

amcasey commented on src/Middleware/Diagnostics/src/PublicAPI.Unshipped.txt:
```diff
@@ -1 +1,3 @@
 #nullable enable
+Microsoft.AspNetCore.Diagnostics.StatusCodeReExecuteFeature.OriginalStatusCode.get -> int
+Microsoft.AspNetCore.Diagnostics.StatusCodeReExecuteFeature.OriginalStatusCode.set -> void
```
```suggestion
```

amcasey committed:
Remove now-internal API from public manifest

amcasey commented:
/azp run
